### PR TITLE
rename .ready archive status files

### PR DESCRIPTION
**Pull request by gsmol**
Hello!
I would like to propose a PR, which introduces the walE-like behavior of  renaming `.ready` status files into `.done` for successfully upladed WAL segments.
It allows to save time by not attempting to upload already uploaded files.
This behaivor is enabled by `--pg-ready-rename` flag.
Feedback and thought on possible improvements are welcome!

### CODE REVIEW

File: internal/databases/postgres/bguploader.go [hunk 1]
```diff
@@ -36,6 +36,7 @@ type BgUploader struct {
 	uploader *WalUploader
 
 	preventWalOverwrite bool
+	ReadyRename         bool
```
Comment (x4m) [inline internal/databases/postgres/bguploader.go:39]: Do we really need it to be public?

File: internal/databases/postgres/bguploader_test.go [hunk 2]
```diff
@@ -98,8 +98,8 @@ func TestBackgroundWALUpload(t *testing.T) {
 			tu := testtools.NewMockWalUploader(false, false)
 			fakeASM := asm.NewFakeASM()
 			tu.ArchiveStatusManager = fakeASM
-			bu := postgres.NewBgUploader(a, int32(tt.maxParallelism), int32(tt.maxNumFilesUploaded), tu, false)
 
+			bu := postgres.NewBgUploader(a, int32(tt.maxParallelism), int32(tt.maxNumFilesUploaded), tu, false, false)
```
Comment (x4m) [inline internal/databases/postgres/bguploader_test.go:102]: BTW, we would need a test for new functionality.

File: internal/databases/postgres/wal_push_handler.go [hunk 3]
```diff
@@ -82,7 +87,24 @@ func uploadWALFile(uploader *WalUploader, walFilePath string, preventWalOverwrit
...
 	}
 	err = uploader.UploadWalFile(walFile)
-	return errors.Wrapf(err, "upload: could not Upload '%s'\n", walFilePath)
+
+	if err != nil {
+		return errors.Wrapf(err, "upload: could not Upload '%s'\n", walFilePath)
+	}
+
+	// rename WAL status file ".ready" to ".done" if requested
+	if ReadyRename && err == nil {
```
Comment (x4m) [inline internal/databases/postgres/wal_push_handler.go:96]: Maybe it's better to do this in ArchiveStatusManager? Or somewhere near...

**Review by x4m — CHANGES_REQUESTED** [review]

File: internal/databases/postgres/bguploader.go (hunk 1 above)
Comment (gsmol) [reply internal/databases/postgres/bguploader.go:39]: nope, my bad

File: internal/databases/postgres/bguploader_test.go (hunk 2 above)
Comment (gsmol) [reply internal/databases/postgres/bguploader_test.go:102]: will do

File: internal/databases/postgres/wal_push_handler.go (hunk 3 above)
Comment (gsmol) [reply internal/databases/postgres/wal_push_handler.go:95]: Good idea, it should be cleaner that way

**Comment by vyegorov** [comment]
I object to such functionality.
This makes it impossible to direct single WAL file to 2 different backup solutions (I use it during migration from one solution to the other). Currently, to make things work for `wal-e`, I'm forced to disable any parallel WAL processing.

**Comment by x4m** [comment]
@vyegorov it will be off by default.

File: internal/databases/postgres/wal_push_handler.go (hunk 3 above)
Comment (gsmol) [reply internal/databases/postgres/wal_push_handler.go]: done

File: internal/databases/postgres/bguploader.go (hunk 1 above)
Comment (gsmol) [reply internal/databases/postgres/bguploader.go]: fixed

File: internal/databases/postgres/bguploader_test.go (hunk 2 above)
Comment (gsmol) [reply internal/databases/postgres/bguploader_test.go:102]: added test

**Review by x4m — APPROVED** [review]
Looks good, thanks!

**Comment by usernamedt** [comment]
Hi!

I found a small issue with docker tests:
1. archiving_ready_rename.sh does not seem to have an executable attribute (chmod +x).
2. there is no entry for this test in[ github actions CI workflow](https://github.com/wal-g/wal-g/blob/01c227fae07fee8e91a2392b3a5b42ddf7900e66/.github/workflows/dockertests.yml#L73), so right now it is not being executed by CI.

**Comment by gsmol** [comment]
@usernamedt, hello!
I will look into it.

**Pull request merged**